src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,152 @@
+            } else {
+                Node<T> newNode = new Node<>(nodeAtIndex.previous, value, nodeAtIndex);
+                newNode.previous.next = newNode;
+                nodeAtIndex.previous = newNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidation(index);
+        return findNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexValidation(index);
+        Node<T> nodeToChange = findNode(index);
+        T oldItem = nodeToChange.item;
+        nodeToChange.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidation(index);
+        Node<T> nodeToRemove = findNode(index);
+        if (size == 1) {
+            head = null;
+            tail = null;
+        } else {
+            if (nodeToRemove == head) {
+                head = nodeToRemove.next;
+                head.previous = null;
+                nodeToRemove.next = null;
+            } else if (nodeToRemove == tail) {
+                tail = nodeToRemove.previous;
+                tail.next = null;
+                nodeToRemove.previous = null;
+            } else {
+                nodeToRemove.next.previous = nodeToRemove.previous;
+                nodeToRemove.previous.next = nodeToRemove.next;
+            }
+        }
+        size--;
+        return nodeToRemove.item;
     }
 
     @Override
     public boolean remove(T t) {
+        Node<T> nodeToFind = head;
+        for (int i = 0; i < size; i++) {
+            if (nodeToFind.item == null ? t == nodeToFind.item : nodeToFind.item.equals(t)) {
+                remove(i);
+                return true;
+            }
+            nodeToFind = head.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findNode(int index) {
+        Node<T> nodeToFind;
+        if (index < (size >> 1)) {
+            nodeToFind = head;
+            for (int i = 0; i < index; i++) {
+                nodeToFind = nodeToFind.next;
+            }
+        } else {
+            nodeToFind = tail;
+            for (int i = size - 1; i > index; i--) {
+                nodeToFind = nodeToFind.previous;
+            }
+        }
+        return nodeToFind;
+    }
+
+    public void indexValidation(int index) {
+        if (index < 0 || index > size - 1) {
+            throw new IndexOutOfBoundsException("Not valid index.");
+        }
+    }
+
+    public void indexValidationForAdd(int index) {
```

Comment:
@TimoxaV we can avoid this method. Check the morning lesson example

Reply:
fixed